lib/target.ml

```diff
@@ -19,9 +19,67 @@ type 'st pre_extern =
 type 'st apply =
   ctrl -> env -> 'st -> signal -> value -> Expression.t option list -> env * 'st * signal * value
 
-type writer = bool -> (string * value) list -> string -> value -> value
+module State = struct
+
+  type 'a t = {
+    externs : (string * 'a) list;
+    heap : (string * value) list;
+    packet : pkt;
+  }
+
+  let empty = {
+    externs = [];
+    heap = [];
+    packet = {emitted = Cstruct.empty; main = Cstruct.empty; in_size = 0; }
+  }
+
+  let packet_location = "__PACKET__"
+
+  let counter = ref 0
+
+  let fresh_loc = fun () ->
+    counter := !counter + 1;
+    "_" ^ (string_of_int (!counter)) ^ "_" 

```

Comment:
Maybe change the "_" to a `"_fresh_loc_"` or something to make it a little more obvious where the strings are coming from if they pop up in error messages